### ISSUE TITLE
Adjust clean up of temporary ingest files

### ISSUE THREAD

**Issue by jrochkind** [sciencehistory/scihist_digicoll#998]
### Original feature was a sketch

1-2 years ago we implemented the "Add from cloud" feature, that allows staff users to first copy assets over to a temporary ingest holding pen S3 bucket, and then select files from that "holding pen" for attaching to an ingest. When it does this, it makes another copy of the file in another location, based off of what was in the temporary holding pen. 

At that time, we considered -- should the file in temporary holding pen be automatically deleted after ingest? 

At that time, we decided against this, in my memory based on "Well, if the ingest fails, we want it still to be there to try again."   Also possibly that it was a bit non-completely-straightforward to implement. 

But we don't want to leave the files on that tempoary holding pen forever, because we'd have to pay for their storage there, and because they'd be clogging up the list of files for staff to choose from!

So, at that time -- before this new feature had ever been used -- we came up with the simplest thing we thought could work. Figuring that we might have to revisit that later. And that was, just leave the things in the holding pen bucket for 7 days after they were added there. The idea was that normally they are added right before they are going to be used.  7 days after it's been added, it gets deleted, the logic doesn't know about when it's actually been ingested, because we were using a simple S3 "lifecycle rule" feature on the S3 buccket. 

More recently we expanded 7 days to 30. 

### Bulk load of batch of Oral Histories to temporary ingest holding pen bucket

Recently (approx Feb 24-26?) Chuck added a giant batch of OH files to the ingest holding pen, so they'd be there for the project work of this current batch, to make ingest as efficient as possible. 

Because we might not be done with them all even in 30 days, our current logic is not necessarily great -- what if they accidentally expire when they aren't all ingested yet? We'd have to do it again. 

It is a bit of a pain/cost to "do it again"; staff needs to find/identify the files to copy to holding pen; Chuck needs to run a routine that can take 24+ hours to complete to copy over the 1.7TB; and it costs about $4.00 (four and no cents) to copy over all of those files. 

While the ~1.7TB of these files exist in the temporary holding pen, we pay about $40/month to store them there, if they are all there without being deleted. (This can not necessarily be used to predict our added costs of indefinite storage in digital collections, as there are things that make the permanent storage both more and less expensive than this temporary holding pen; although it can be a rough order of magnitude. This $40 is for an *extra* copy that is just sitting in the temporary holdingpen). 

## What should we do?

The current implementation of "just delete it in X days without knowing if it's been ingested" was always a sort of "kick the can down the road solution", we weren't sure exactly what to do, so just did the simplest cheapest thing, knowing we might revisit later. 

It might be time to revisit? 

### Do not revisit, just turn off the 30 day limit. 

Rather than make it longer, we'd probably just turn it off altogether for now. We'd plan on later, once we're all done with this batch of ingests, delete *all* the files in the temporary holding pen, and then maybe turn back on the 30 day limit. 

* We have no easy way to tell which of the files in the temporary ingest holding pen are part of this special OH batch, and which might be other files, so we'd really have to delete *all* files, possibly interupting other work, then re-add any others that should be there. 
* We might be paying more for storage than needed, if we're paying $40 a month, at the point we are 50% done with ingest, $20 worth of those files *could* have been deleted because we are done with them, but we're still paying for them. 
* While we are saving developer time by not writing new functionality, we might lose some by having to manually try to wrangle this large number/size of files. Having to remember to go back and do it etc. If we forget, accidentally paying money for S3 storage we didn't need to be. 
* Even under normal circumstances, the 30 day limit means we're paying more money than we need to, maybe keeping things around for 30 days that don't actually need to be there, they've already been ingested.  It's not a huge amount of money, but can add up if we keep doing increasingly large volumes of ingest. As noted, around $40/month for the 1.7TB in this OH batch. 

Would different logic be more convenient not just for OH project where it's really coming to a head, but in general? This was originally intended as a "kick the can down the road" "simplest possible solution", if we do some work to make this work better for current workflow needs, will it pay dividends even beyond OH ingest project?

### Automatically delete on ingest?

Are we confident enough in the Digital Collections software to just say "delete all the files from temporary holding pen right after they've been ingested"?

How often have staff needed to go back and re-ingest?  If this is rare, is it a big deal if they just need to copy the file to s3 *again* when this happens?

This can probably be implemented relatively easy, maybe 1-3 days of work. 

Is it a problem that if we add something to temporary holding pen and accidentally never get around to ingesting it, it will just stay there forever with us paying storage costs? We'd just count on staff noticing that, and bringing it to dev attention to delete the files, I gyess?

### Automatically delete X days after ingest

Where X is maybe 7. 

Current functionality is to delete X days after *adding to temporary holding pen*. 

It might make more sense to automatically delete X days after *actually ingesting*. Then it's still there "just in case ingest didn't work" for 7 days, but it's deletion is more tied to "when we actually got around to ingesting it". 

Possibly still issue with adding something to temporary holding pen and accidentally never get around to ingesting it. 

This will be a bit harder to implement but still not too bad, maybe a week-ish. 

### UI for manually deleting things

Right now developers can manually delete files manually selected using the AWS Console, but ingest staff has no way to do so. 

We could create a UI to let them choose files on temporary holding pen to delete. 

I don't think this is very useful as a solution, we don't want to add *another* step to ingest process,.  After ingesting, staff would have to go to this other UI and choose files to delete when they are ready to delete them from temporary holding pen S3 bucket. This adds an annoying step to ingest process and is error prone (someone could forget to delete; or could delete too early when the file is still needed). 

I don't think it's a good solution, just listing it for completeness.

**Comment by MDiMeo** [comment]
Talked to Hillary about the workflow and weighted pros and cons. We have decided on Option 3: Automatically delete 7 days after ingest

**Comment by jrochkind** [comment]
I manually deleted the "delete 30 days after upload" rule in AWS console from the bucket `scihist-digicoll-production-ingest-mount`. 

i will also PR to our ansible-inventory playbook to remove it from the ansible playbook for creating S3. 

I tested, and removing it from the playbook and re-running it is not alone enough to remove it -- the playbook is smart enough to add it if it's not there, but not to remove it once we've removed it from our playbook. One reason ansible might not be the ideal tool for trying to maintain S3 configuration-as-code.
